src/python/WMComponent/DBS3Buffer/DBSUploadPoller.py

```diff
@@ -718,16 +718,28 @@ def retrieveBlocks(self):
                 updateBlocksDAO.execute(blocks = loadedBlocks,
                                         conn = myThread.transaction.conn,
                                         transaction = True)
-            except WMException:
-                myThread.transaction.rollback()
-                raise
             except Exception as ex:
                 myThread.transaction.rollback()
-                msg =  "Unhandled exception while finished closed blocks in DBSBuffer\n"
-                msg += str(ex)
-                logging.error(msg)
-                logging.debug("Blocks for Update: %s\n" % loadedBlocks)
-                raise DBSUploadException(msg)
+                # possible deadlock with PhEDExInjector, retry once after 10s
+                time.sleep(10)
+                try:
+                    myThread.transaction.begin()
+                    updateFilesDAO.execute(blocks = loadedBlocks, status = "InDBS",
+                                           conn = myThread.transaction.conn,
+                                           transaction = True)
+                    updateBlocksDAO.execute(blocks = loadedBlocks,
+                                            conn = myThread.transaction.conn,
+                                            transaction = True)
+                except Exception as ex:
+                    myThread.transaction.rollback()
+                    msg =  "Unhandled exception while finished closed blocks in DBSBuffer\n"
+                    msg += str(ex)
+                    logging.error(msg)
+                    logging.debug("Blocks for Update: %s\n", loadedBlocks)
+                    raise DBSUploadException(msg)
+                else:
+                    myThread.transaction.commit()
```

Comment:
Well, that whole section of code is one big duplication because we do the retry if it fails in the exception. I rather duplicate the whole blocks of code here because this is ugly to start with, rather show the complete functional blocks to show what I am doing.

Another reason I want to leave it in is that the else here isn't needed due to the raise in the except. But there can be cases where you don't raise (just log the error and move on), then you absolutely need the else.

I am using this as a general purporse try/except/else construct everywhere in my code.